dbms/src/DataStreams/WindowTransformAction.h

```diff
@@ -160,6 +155,77 @@ struct WindowTransformAction
     // distance is left - right.
     UInt64 distance(RowNumber left, RowNumber right);
 
+    void initialWorkspaces();
+    void initialPartitionAndOrderColumnIndices();
+    void initialAggregateFunction(
+        WindowFunctionWorkspace & workspace,
+        const WindowFunctionDescription & window_function_description);
+
+    void addAggregationState(WindowFunctionWorkspace & ws, const RowNumber & start, const RowNumber & end)
+    {
+        addOrDecreaseAggregationState<true>(ws, start, end);
+    }
+
+    void decreaseAggregationState(WindowFunctionWorkspace & ws, const RowNumber & start, const RowNumber & end)
+    {
+        addOrDecreaseAggregationState<false>(ws, start, end);
+    }
+
+    template <bool is_add>
+    void addOrDecreaseAggregationState(WindowFunctionWorkspace & ws, const RowNumber & start, const RowNumber & end)
```

Comment:
Seems there is no need to define this function in header file?

Reply:
> Seems there is no need to define this function in header file?

Defition and implementation are better to be put together or we will need to manually initialize it.

Reply:
I don't get your point, will we need initialize a function?